lua/entities/gmod_tardis/modules/sh_health.lua

```diff
@@ -0,0 +1,174 @@
+    cvars.AddChangeCallback("tardisrw_maxhealth", function(cvname, oldvalue, newvalue)
+       TARDIS:SetSetting("health-max", tonumber(newvalue), true)
+    end, "UpdateOnChange")  
+    
+    cvars.AddChangeCallback("tardisrw_damage", function(cvname, oldvalue, newvalue)
+       TARDIS:SetSetting("health-enabled", tobool(newvalue), true)
+    end, "UpdateOnChange")
+
+    function ENT:Explode()
+        if not vFireInstalled then
+            local explode = ents.Create("env_explosion")
+            explode:SetPos( self:LocalToWorld(Vector(0,0,50)) ) //Puts the explosion where you are aiming
+            explode:SetOwner( self ) //Sets the owner of the explosion
+            explode:Spawn()
+            explode:SetKeyValue("iMagnitude","100") //Sets the magnitude of the explosion
+            explode:Fire("Explode", 0, 0 ) //Tells the explode entity to explode
+            //explode:EmitSound("weapon_AWP.Single", 400, 400 ) //Adds sound to the explosion
+        else
+            //alternate explosion
+        end
+    end
+
+    function ENT:ToggleRepair()
+        local on = not self:GetData("repair-primed",false)
+        self:SetRepair(on)
+    end
+    function ENT:SetRepair(on)
+        //self.interior:SetData("selfrepair-primed",on,true)
+        if (self:GetHealth() > TARDIS:GetSetting("health-max",1)-1) or self:GetData("vortex",false) then return end
+        if on==true then
+            for k,_ in pairs(self.occupants) do
+                k:ChatPrint("This TARDIS has been set to self-repair. Please vacate the interior.")
+            end
+            if self.interior:GetData("power-state") then self.interior:SetPower(!on) end
+            self:SetData("repair-primed",on,true)
+        else
+            self:SetData("repair-primed",on,true)
+            self.interior:SetPower(!on)
+            for k,_ in pairs(self.occupants) do
+                k:ChatPrint("TARDIS self-repair has been cancelled.")
+            end
+        end
+    end
+
+    function ENT:FinishRepair()
+        self:EmitSound(self.metadata.Exterior.Sounds.RepairFinish)
+        self:SetData("repairing", false, true)
+        self:ChangeHealth(TARDIS:GetSetting("health-max"))
+        self.interior:SetPower(true)
+        self:SetLocked(false, nil, true)
+        self:GetCreator():ChatPrint("Your TARDIS has finished self-repairing")
+        self:FlashLight(1.5)
+    end
+
+    ENT:AddHook("CanTogglePower", "health", function(self)
+        if (self:GetData("health-val", 0) > 0) and (self:GetData("repairing",false)==false and self:GetData("repair-primed", false)==false) then
+            return true
+        end
+    end)
+
+    ENT:AddHook("CanLock", "repair", function(self)
+        if (not self:GetData("repairing",false)) then return true end
+    end)
+
+    ENT:AddHook("PlayerExit", "repair", function(self,ply,forced,notp)
+        if (self:GetData("repair-primed",false)==true) and (#self.occupants==0) then
+            timer.Simple(0.3, function()
```

Comment:
What if the last player excited and then entered really quickly again? Maybe check again to ensure no occupants or remove the timer entirely? I discourage use of timer.Simple for reasons like this

Reply:
Also what if the tardis has been deleted in between this time? CurTime checks may be better here

Reply:
I couldn't think of a solution for that at the time and kind of forgot about it. Thanks for the reminder :p